src/ai/backend/manager/server.py

```diff
@@ -495,6 +496,122 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    ) -> tuple[SessionRow, ...]:
+        query = (
+            sa.select(SessionRow)
+            .where(SessionRow.status == status)
+            .where(
+                (
+                    datetime.now(tz=tzutc())
+                    - SessionRow.status_history[status.name].astext.cast(
+                        sa.types.DateTime(timezone=True)
+                    )
+                )
+                > threshold
+            )
+            .options(
+                noload("*"),
+                load_only(SessionRow.id, SessionRow.name, SessionRow.status, SessionRow.access_key),
+            )
+        )
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            return result.fetchall()
+
+    async def _force_terminate_hanging_sessions(
+        status: SessionStatus,
+        threshold: timedelta,
+        interval: float,
+    ) -> None:
+        try:
+            sessions = await _fetch_hanging_sessions(
+                root_ctx.db, status=status, threshold=threshold
+            )
+        except asyncio.CancelledError:
+            pass
+        except Exception as e:
+            log.error("fetching hanging sessions error: {}", repr(e), exc_info=e)
+            return
+
+        log.debug(f"{len(sessions)} {status.name} sessions found.")
+
+        results_and_exceptions = await asyncio.gather(
+            *[
+                asyncio.create_task(
+                    root_ctx.registry.destroy_session(
+                        session, forced=True, reason=KernelLifecycleEventReason.HANG_TIMEOUT
+                    ),
+                )
+                for session in sessions
+            ],
+            return_exceptions=True,
+        )
+        for result_or_exception in results_and_exceptions:
+            if isinstance(result_or_exception, (BaseException, Exception)):
+                log.error(
+                    "hanging session force-termination error: {}",
+                    repr(result_or_exception),
+                    exc_info=result_or_exception,
+                )
+
+    hang_tolerance_threshold_dict = session_hang_tolerance_threshold_iv.check(
+        await root_ctx.shared_config.etcd.get_prefix_dict("config/session/hang-tolerance-threshold")
+    )
+
+    session_force_termination_tasks = []
+    heuristic_interval_weight = 0.4  # NOTE: Shorter than a half(0.5)
+    for status, threshold in hang_tolerance_threshold_dict.items():
+        try:
+            session_status = SessionStatus[status]
+        except KeyError:
+            continue
+
+        if isinstance(threshold, relativedelta):  # months, years
+            if months := threshold.months:
+                threshold = timedelta(seconds=months * 30 * 24 * 60 * 60)
+            elif years := threshold.years:
+                threshold = timedelta(seconds=years * 365 * 24 * 60 * 60)
+            else:
+                log.warning("Skipping {}:{}...", status, threshold)
+                continue
```

Comment:
`relativedelta` considers leap years and the difference of the number of days in months, while `timedelta` is a single absolute value.
So, we should keep the original form of `relativedelta` until we do the actual comparison instead of converting it to `total_seconds`.

See https://stackoverflow.com/a/49021582.
Could we move this into the timer function?

Reply:
Note that if the user's intention is to specify an absolute value of time difference, still one can achieve the goal using a single-character suffixes when specifying the threshold value (e.g., `365d` instead of `1yr`).